source/MRMesh/MRCylinderApproximator.h

```diff
@@ -0,0 +1,373 @@
+    std::vector<Eigen::Vector<T, 3>> normalizedPoints_ = {};
+
+    // Precalculated values for speed up. 
+    // In https://www.geometrictools.com/Documentation/LeastSquaresFitting.pdf page 35-36: 
+    // Text below is a direct copy from pdf file: 
+    // The sample application that used equation (94) directly was really slow. 
+    // On an Intel CoreTM i7-6700 CPU at 3.40 GHz, the single - threaded version for 10765 points required 129 seconds 
+    // and the multithreaded version using 8 hyperthreads required 22 seconds.The evaluation of G using the precomputed summations is much faster.
+    // The single - threaded version required 85 milliseconds and the multithreaded version using 8 hyperthreads required 22 milliseconds.
+
+    Eigen::Vector <T, 6>  precomputedMu_ = {};
+    Eigen::Matrix <T, 3, 3>  precomputedF0_ = {};
+    Eigen::Matrix <T, 3, 6>  precomputedF1_ = {};
+    Eigen::Matrix <T, 6, 6>  precomputedF2_ = {};
+
+public:
+    Cylinder3Approximation()
+    {
+        reset();
+    };
+
+    void reset()
+    {
+        thetaResolution_ = 0;
+        phiResolution_ = 0;
+        precomputedMu_.setZero();
+        precomputedF0_.setZero();
+        precomputedF1_.setZero();
+        precomputedF2_.setZero();
+        normalizedPoints_.clear();
+    };
+    // Solver for CylinderFitterType::HemisphereSearchFit type 
+    // thetaResolution_, phiResolution_  must be positive and as large as it posible. Price is CPU time. (~ 100 gives good results).
+    T solveGeneral( const std::vector<MR::Vector3<T>>& points, Cylinder3<T>& cylinder, size_t theta = 180, size_t phi = 90, bool isMultithread = true )
+    {
+        thetaResolution_ = theta;
+        phiResolution_ = phi;
+        isMultithread_ = isMultithread;
+        fitter_ = CylinderFitterType::HemisphereSearchFit;
+        assert( thetaResolution_ > 0 );
+        assert( phiResolution_ > 0 );
+        auto result = solve( points, cylinder );
+        reset();
+        return result;
+    };
+
+    // Solver for CylinderFitterType::SpecificAxisFit type 
+    // Simplet way in case of we already know clinder axis 
+    T solveSpecificAxis( const std::vector<MR::Vector3<T>>& points, Cylinder3<T>& cylinder, MR::Vector3<T> const& cylinderAxis )
+    {
+
+        baseCylinderAxis_ = MR::toEigen( cylinderAxis.normalized() );
+        fitter_ = CylinderFitterType::SpecificAxisFit;
+        assert( baseCylinderAxis_.isZero() == false ); //  "The cylinder axis must be nonzero."
+        auto result = solve( points, cylinder );
+        reset();
+        return result;
+    };
+private:
+    // main solver. 
+    T solve( const std::vector<MR::Vector3<T>>& points, Cylinder3<T>& cylinder )
+    {
+        assert( points.size() >= 6 ); // "At least 6 points needs for correct fitting requires ."
+
+        normalizedPoints_.clear();
+        cylinder = Cylinder3<T>();
+        Vector3<T> avgPoint = {};
```

Comment:
no need in `= {}`, it does not change anything here

Reply:
done